### PR TITLE
Use async networking in hub listener

### PR DESCRIPTION
**Pull request by psiborg112**
Refactored the UdpHubListener to use asynchronous networking. (Should prevent a rogue client from blocking subsequent connections by not sending its port.) This is the last of the Qt socket code to be restructured in this way: the only synchronous networking code is now the bsd/winsock code in UdpDataProtocol.

This change also means that UdpHubListener doesn't require its own runloop, so it is no longer derived from QThread. (SIGINT/SIGTERM are monitored using a timer.)

### CODE REVIEW

**Comment by psiborg112** [comment]
@mikedickey I know you've done a fair bit of work testing concurrent hub server connections, so if you're able to run this change through its paces with your set of tests that would be greatly appreciated. (And will hopefully show up any issues.)

**Comment by ntonnaett** [comment]
I would really like to see this in dev. @mikedickey Could you test this?

**Comment by mikedickey** [comment]
Sorry, I'm a bit swamped right now and may not be able to test anything for a few more weeks.

**Comment by psiborg112** [comment]
All good Mike, I know the feeling.

**Comment by josmithiii** [comment]
Should this go before or after Anton's jitter buffer merge?

**Comment by psiborg112** [comment]
I think either order should be fine, but it might be easier to merge 
this first.
Aaron

**Comment by josmithiii** [comment]
I created a branch jb-async consisting of the latest arunov/jitter-buffer with async-hublistener merged in, taking the default merge-conflict resolution in opendiff (only UdpHubListener.cpp), and it built smoothly.  All my JackTrip client tests pass.

**Comment by mikedickey** [comment]
@josmithiii did you push that somewhere? I'm getting a pretty gnarly conflict in UdpHubListener.cpp when I try to merge these together. At least it's too scary for me to try tackling. Didn't seem to matter if I merged this or jitter-buffer first.

I've been using jitter-buffer successfully for a while, so I suggest we merge that to dev first, fix conflicts, then re-test and merge this one.

**Comment by antonrunov** [comment]
I looked through the diff breifly a few days ago. It all looks reasonable but I need to go a bit deeper to say something definitive. And yes, I had some concerns regarding merging with jitter buffer. It should be absolutely solvable, just a couple of places that should be checked though. Will take a closer look soon.

File: src/UdpHubListener.cpp
Comment (antonrunov) [inline src/UdpHubListener.cpp:228]: When merging with JB branch, the new JB related initialization commands from old `UdpHubListener::run` should be moved here.

**Comment by josmithiii** [comment]
> @josmithiii did you push that somewhere? I'm getting a pretty gnarly conflict in UdpHubListener.cpp when I try to merge these together. At least it's too scary for me to try tackling. Didn't seem to matter if I merged this or jitter-buffer first.

I pushed it just now: Branch ```jb-async``` ("jitter buffer + async networking").

I didn't push it earlier because it was an automatic mix by opendiff (which very often does get everything right) - ideally Aaron and/or Anton should check this merged branch (or prepare it from scratch) for testing. 

```On branch jb-sync:
> git diff async-hublistener | wc
    4821   24015  201992
> git diff arunov/jitter-buffer | wc
     369    1852   16636
> git blame UdpHubListener.cpp | grep Aaron | wc
     179    2189   25507
> git blame UdpHubListener.cpp | grep Anton | wc
      13     128    1777```

Thus, most of the changes in this file are Aaron's.

I will study the file myself and see how much I understand.

**Comment by antonrunov** [comment]
Hmm... I don't see Aaron's changes in `jb-async`
```git diff arunov/jitter-buffer jb-async```
shows a single line difference which is actually the last fix to jitter-buffer I made recently.

**Comment by josmithiii** [comment]
> Hmm... I don't see Aaron's changes in `jb-async`
> `git diff arunov/jitter-buffer jb-async`
> shows a single line difference which is actually the last fix to jitter-buffer I made recently.

Yes, that was my re-merge of your branch to get that one line --- an "update commit" if you will.

You can see the diffs this way:
```
> git diff arunov/jitter-buffer async-hublistener UdpHubListener.h
> git diff arunov/jitter-buffer async-hublistener UdpHubListener.cpp
```
Sourcetree is my favorite tool for perusing diffs associated with various past commits in any branch.

**Comment by josmithiii** [comment]
or maybe you'd rather see```
> git diff arunov/jitter-buffer jb-async UdpHubListener.h
> git diff arunov/jitter-buffer jb-async UdpHubListener.cpp
```

**Comment by antonrunov** [comment]
Yes now I see the changes and the merge is fine.

File: src/UdpHubListener.cpp
Comment (antonrunov) [inline src/UdpHubListener.cpp:161]: This is sort of a hack from the viewpoint of Qt signal-slot mechanism. I think it's totally Ok here, but it could be avoided with connecting lambda expression instead of slot in `receivedNewConnection`.

**Comment by antonrunov** [comment]
I think this PR is fine to merge. Julius's branch provides a clear merge with JB branch.

**Comment by josmithiii** [comment]
In that case, the merges are easy: (1) [squash]-merge Anton's PR into dev, (2) merge dev into async-hublistener, taking the default opendiff resolution, (3) commit and push that, which updates Aaron's PR, then finally (4) merge Aaron's PR into dev.  The result should be dev identical to branch jb-async.

File: src/UdpHubListener.cpp
Comment (psiborg112) [reply src/UdpHubListener.cpp:161]: I don't really have a strong opinion either way, but a lambda might be the better option, and it's a fairly trivial fix. I could upload a change now or I could wait until I merge more of my developments over.

Comment (antonrunov) [reply src/UdpHubListener.cpp:161]: I don't have a strong opinion either. It's totally up to you. I just wanted to mention this option hoping it might be helpful.

**Comment by mikedickey** [comment]
I've merged jitter-buffer to dev. Please update this for conflicts, and then I'll re-run my scale tests on it.

**Comment by josmithiii** [comment]
I verified just now that merging async-hublistener into dev reproduces jb-async exactly:
```
git checkout dev
git pull
git merge async-hublistener
;; merge conflicts noted
git mergetool
;; take default opendiff merge, i.e., just save and quit
git commit -a -m 'merge async-hublistener into dev'
git diff jb-async
;; no differences
```

**Comment by mikedickey** [comment]
Just checking back to see if we can get this PR updated so that I can test.

Do we still want to try to include this in a new release? Or should we consider cutting a 1.3.0 release off current dev branch and then merge this after?

**Comment by psiborg112** [comment]
We should definitely include this because it prevents an incredibly easy to execute denial of service attack. (Someone can simply open a TCP connection to a jacktrip server, not send anything, and no one else will be able to connect.) I just won't have time to properly update it in the next few days.

**Comment by psiborg112** [comment]
Ok, had a rehearsal finish early, so managed to quickly get this done. Bit of a rush job though, so someone should definitely double check that I haven't missed anything.

**Review by mikedickey — APPROVED** [review]
Tested with up to 481 clients: same results as dev

**Pull request merged**
